compiler/rustc_hir_typeck/src/op.rs

```diff
@@ -380,33 +380,95 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                                "`{}{}` can be used on `{}` if you dereference the left-hand side",
+                                op.node.as_str(),
+                                match is_assign {
+                                    IsAssign::Yes => "=",
+                                    IsAssign::No => "",
+                                },
+                                lhs_deref_ty,
+                            );
+                            err.span_suggestion_verbose(
+                                lhs_expr.span.shrink_to_lo(),
+                                msg,
+                                "*",
+                                rustc_errors::Applicability::MachineApplicable,
+                            );
+                        }
+                    };
+
+                let suggest_different_borrow =
+                    |err: &mut DiagnosticBuilder<'_, _>,
+                     lhs_adjusted_ty,
+                     lhs_new_mutbl,
+                     rhs_adjusted_ty,
+                     rhs_new_mutbl| {
+                        if self
+                            .lookup_op_method(
+                                lhs_adjusted_ty,
+                                Some((rhs_expr, rhs_adjusted_ty)),
+                                Op::Binary(op, is_assign),
+                                expected,
+                            )
+                            .is_ok()
+                        {
+                            let op_str = op.node.as_str();
+                            err.note(format!("an implementation for `{lhs_adjusted_ty} {op_str} {rhs_adjusted_ty}` exist"));
+
+                            let mut suggest_new_borrow = |new_mutbl, sp: Span| {
+                                // Can reborrow (&mut -> &)
+                                if new_mutbl == ast::Mutability::Not {
+                                    err.span_suggestion_verbose(
+                                        sp.shrink_to_lo(),
+                                        "consider reborrowing this side",
+                                        "&*",
+                                        rustc_errors::Applicability::MachineApplicable,
+                                    );
+                                // Works on &mut but have &
+                                } else {
+                                    err.span_help(
+                                        sp,
+                                        "consider making this expression a mutable borrow",
+                                    );
+                                }
+                            };
+
+                            if let Some(lhs_new_mutbl) = lhs_new_mutbl
+                                && let Some(rhs_new_mutbl) = rhs_new_mutbl
+                                && lhs_new_mutbl == ast::Mutability::Not
+                                && rhs_new_mutbl == ast::Mutability::Not {
+                                drop(suggest_new_borrow);
```

Comment:
I think it makes more sense to move the closure definition in the else branch.

Reply:
Good call, wasn't possible before. Done.